New Web Monetization Specification Proposal

marcoscaceres opened the pull request:
A preview of the new proposed specification can be viewed at https://webmonetization-preview.netlify.app/

Removed:
  * using meta elements
  * various conformance restrictions 
  * flow part... will be part of link relationship instead
  * state entirely
  * document.monetization
 
Added: 

* Allow multiple link elements - but the WM agent can limit how many connections it handles.
* The "progress" event is now called "monetization", and is fired at the <link> element .
* introduce a new global event handler attribute: `onmonetization`, which is on every HTML element.
* We now route all events through the <link> element.
* Developer-recoverable errors are handled via the `<link onerror>` event handler.
* Successfully loading/processing SPSP is reflected via the `<link onload>` event handler.
* Defined CSP `monetization-src`. 
* assetScale in now an octet (uint8).

Closes #144
Closes #149
Closes #159  
Closes #19
Closes #192
Closes #20
Closes #201
Closes #21 
Closes #214
Closes #215
Closes #218
Closes #232
Closes #22 
Closes #233
Closes #28
Closes #29
Closes #41
Closes #23
Closes #216
Closes #220 
Closes #165 
Closes #204 
Closes #219  
Closes #237

sharafian commented on static/specification.html:
```diff
@@ -59,384 +58,130 @@
+        </dt>
+        <dd>
+          The party making payments on behalf of the user. A Web Monetization
+          provider can, for example, make use of the Simple Payment Setup
```
This wording should be changed because making use of SPSP on top of ILP is mandated by the spec rather than being one option of several

sharafian commented on static/specification.html:
```diff
@@ -445,42 +190,52 @@ <h2><dfn>Monetization</dfn> interface</h2>
+        </section>
+        <section>
+          <h2>
+            <dfn>onstart</dfn> attribute
```
do we still need the start event, or could the first progress event just be treated as a start event by the page?

marcoscaceres replied:
It would be nice if start went away. You can check the state independently.

sharafian commented on static/specification.html:
```diff
@@ -514,210 +271,189 @@ <h2>
-              readonly attribute DOMString paymentPointer;
+            interface MonetizationEvent : Event {
+              constructor(DOMString type, MonetizationEventInit eventInitDict);
+              readonly attribute USVString paymentPointer;
```
if we're no longer using the payment pointer format and using URL format instead, it might be better to call this something like `receiver`

AlexLakatos replied:
Would this be the ILP address, or the SPSP endpoint?

marcoscaceres replied:
Ok, which one is it going to be? 🙏

adrianhopebailie replied:
@sharafian isn't it still a payment pointer even if it's in the URL form?

marcoscaceres replied:
Related https://github.com/WICG/webmonetization/issues/201

marcoscaceres replied:
I think I'm going to switch this to `url`. It's also what @sidvishnoi did in his implementation.

sublimator replied:
> if we're no longer using the payment pointer format and using URL format instead, it might be better to call this something like receiver

sublimator replied:
Payment pointers are referenced extensively in Open API   
https://github.com/interledger/open-payments/blob/main/open-api-spec.yaml

sharafian commented on static/specification.html:
```diff
@@ -514,210 +271,189 @@ <h2>
+              dictionary MonetizationEventInit : EventInit {
+                required USVString paymentPointer;
+                required DOMString requestId;
+                DOMString amount;
```
how come amount isn' required?

marcoscaceres replied:
This was just to cater to the "start" event, when you might not have an amount.  However. the events are created by the (trusted) WM agent, so technically we can assure that `amount` will always be there (same with the rest).

AlexLakatos replied:
If this is rolled into one event, they'll all look the same, and all the optional aren't optional anymore.

sharafian commented on static/specification.html:
```diff
@@ -59,384 +58,130 @@
+        monetize the document.
+      </p>
+      <p>
+        ...Fetching and processing details TBW...
```
What sort of changes are we anticipating to the flow?

marcoscaceres replied:
I'm basically planning on transplanting: https://html.spec.whatwg.org/#link-type-manifest and adapting it to this. 

That part of HTML handles the fetching and processing and eventually will call back into here (where anything "flow" specific will take over). 

Other changes will be just hooking into the DOM/HTML eventing model. I'll try to have a draft of that this coming week or next.

sublimator replied:
I am not sure that the manifest link type actually matches how we use the tag
* We request the resource (payment endpoint) more than once per document load
* We sometimes do not even request the resource, rather send it to the backend to be used in a potentially async manner. 
* This makes assumptions about how how much the browser will actually implement vs any potential WM Provider plugins ( perhaps implemented as extensions as per @sidvishnoi ) that may want access, for whatever reason, to the payment endpoint url. 

I think we want something closer to a dumber (metadata) tag, like "<link rel='author' >"  that simply encodes a relationship, rather than something that implies a (read: one) fetch,  as per  https://html.spec.whatwg.org/#external-resource-link

> * Links to external resources
These are links to resources that are to be used to augment the current document, generally automatically processed by the user agent. All [external resource links](https://html.spec.whatwg.org/#external-resource-link) have a [fetch and process the linked resource](https://html.spec.whatwg.org/#fetch-and-process-the-linked-resource) algorithm which describes how the resource is obtained.

adrianhopebailie replied:
@sublimator - The choice of a `link` tag was very intentional based on feedback from browser vendors and is supposed to mimic the behaviour of other linked resources that are fetched.

I think we need to review the lifecycle of the link and the events that are emitted to make sure we're all on the same page as to what is intended and what developers expect.

Maybe the right way to think about this is to have two separate events and be very clear to developers what these mean:
1. Load - the browser has successfully fetched the payment pointer and was able to do whatever processing is necessary to start sending payments if the WM provider wants to send payments. This event will not fire if the user has disabled monetization either globally or for this origin 
2. Monetization - a payment was sent to the payment pointer

sublimator replied:
To be clear, I'm not against a link tag, but the way we are using it. 

At this point it seems completely obvious to some (I think you and Alex are on the same page here *) that the Browser should be the only component ever to have access to the payment pointer. Only the processed link connection details should be shared with other components. 

Uncompromising on privacy, baking in the tradeoffs.  

With that, sure, why not a 'load' event?

If we are going to have a 'load' event, and it means processing the payment pointer, could we emit it as early as possible? As soon as the payment pointer seems well formed, and the connection details requests are about to be initiated? 

I have a bad habit of jumbling in a lot of different points, but here we go: 
I once asked why we (Coil) don't just do away with WebSockets/BTP and set up a simple http payment endpoint, that could be used for batch payments and tipping, authenticated with privacy pass tokens. Apparently, with the amounts involved with tipping, KYC/AML is more of an issue. That answer could/should be revisited, but I think it's worth considering.

sublimator replied:
> If we are going to have a 'load' event, and it means processing the payment pointer, could we emit it as early as possible? As soon as the payment pointer seems well formed, and the connection details requests are about to be initiated?

I think that would:
1.  Make it future compatible with sharing the payment pointer outside of the Browser/UA if that tradeoff is ever revisited 
2.  Make it so that websites don't have to wait precious milliseconds before making decisions based on the load event, (which does NOT necessarily guarantee payment anyway)

adrianhopebailie replied:
> If we are going to have a 'load' event, and it means processing the payment pointer, could we emit it as early as possible? As soon as the payment pointer seems well formed, and the connection details requests are about to be initiated?

I don't see why not.

sublimator replied:
> The choice of a link tag was very intentional based on feedback from browser vendors

It's interesting. I also talked to people on the Chrome payments team who said that meta tags might actually be better. I guess it really depends how you ask the questions.

marcoscaceres commented:
Thanks for the feedback @sharafian! Will update soon. 

I'll break this up and start sending the editorial changes first.

AlexLakatos commented on static/specification.html:
```diff
@@ -59,384 +58,130 @@
+          <dfn>Payment end-point</dfn>:
+        </dt>
+        <dd>
+          A [=URL=] to an [[[OpenPayments]]] conforming HTTP end-point.
```
That's an ["SPSP Endpoint"](https://interledger.org/rfcs/0009-simple-payment-setup-protocol/#specification) according to the ILP spec. Maybe we should change the Open Payments mention to SPSP.

AlexLakatos commented on static/specification.html:
```diff
@@ -59,384 +58,130 @@
-        <li>
-          Should be compatible with existing application and transport
-          protocols, such as the [[[Interledger]]].
+        <li>Compatible with existing monetization protocols, such as the
```
Technically, Interledger is a "protocol for sending packets of money across different payment networks or ledgers". I think monetization is a use case of it, but not a category.

AlexLakatos commented on static/specification.html:
```diff
@@ -445,42 +190,52 @@ <h2><dfn>Monetization</dfn> interface</h2>
+            <dfn>onprogress</dfn> attribute
+          </h2>
+          <p>
+            A {{Monetization/onprogress}} attribute is an A
```
This reads a bit confusing on the generated document. I think removing this line entirely would help.

`A onprogress attribute is an A onprogress attribute is an EventHandler for an MonetizationEvent named "progress". `

marcoscaceres replied:
Yeah, sorry, something got lost here... fixing.

AlexLakatos reviewed (changes requested):
Looks good overall, just a few technicalities in there from the Interledger terminology.

marcoscaceres commented:
Started sending the various parts as pull requests... eventually this will end up with just the link relationship.

sidvishnoi commented on static/specification.html:
```diff
@@ -144,289 +144,296 @@ <h2>
+        controls the monetization. It is initialized as `null`.
+      </p>
+      <p>
+        The <code data-x="rel-monetization">monetization</code> keyword may be
```
What is this `data-x` thing?

marcoscaceres replied:
This is watsi. This will eventually move to the HTML spec.

sidvishnoi commented on static/specification.html:
```diff
@@ -144,289 +144,296 @@ <h2>
+        <ul>
+          <li>
+            <p>
+              When {{Navigator}}'s {{Navigator\[[activeMonetizationElement]]}}
```
```suggestion
              When {{Navigator}}'s {{Navigator/[[activeMonetizationElement]]}}
```

sidvishnoi reviewed (commented):
Some questions on ReSpec usage..

Also, here's the live spec as pr-preview seems to be not working:
https://raw.githack.com/WICG/webmonetization/link_rel/static/specification.html

sublimator commented:
iic there was some talk prior of maybe merging amount/assetCode/assetScale into a single field? 
What was the thoughts on that ?

marcoscaceres commented:
> amount/assetCode/assetScale into a single field?
What were the thoughts on that ?

Had the same question. That be amazing. New issue?

sublimator commented on static/specification.html:
```diff
@@ -410,11 +410,10 @@ <h2>
           |element| and the following steps:
             <ol>
-              <li>If |success| is true, [=fire an event=] named "load" at
-              |element|, which bubbles.
```
@marcoscaceres 
FYI: 
![image](https://user-images.githubusercontent.com/525211/145168409-43628067-0164-43e6-a9ea-68aaa8c7345b.png)
Url out of date: 
https://webmonetization-preview.netlify.app/#link-type-monetization

marcoscaceres replied:
yeah, I need to set up a solution to generate the preview... bugging @AlexLakatos every time I make a change is not going to scale 😝

sidvishnoi replied:
Maybe can use https://wicg.github.io/webmonetization for spec preview and https://webmonetization.org for all the things?

marcoscaceres replied:
yeah, something like that would work...

AlexLakatos replied:
In the meantime, I've re-deployed, so now the preview is up to date.

sentientwaffle commented:
-    §1.1: The example shows `onmonetization="sayThanks()"` and `function sayThanks(event) {` — where does `event` come from?
-    §8, Is `assetScale` intentionally a string (instead of a integer/number)? (If so, why?)
-    Also, why expose delivery amount/units but not sent amount/units?
-    §8.1, typo: `The amount received as by an ILP packet.`. Also, "delivered" might be more appropriate than "received", since this is mostly written from the perspective of the web browser, not the monetization receiver.
-    §8.2, nit: replace "describing" with "identifying" or "indicating".
-    §9, typo: there's an extra double-quote after `monetization`.
- It would be good to either include or link to some more comprehensive examples, e.g. bypassing a paywall.

sublimator commented:
> §8, Is assetScale intentionally a string (instead of a integer/number)? (If so, why?)

Seems like the (WM) spec has always had it (errantly) as an assetScale

> where does `event` come from?

Seems `event` is set on window and is ~~deprecated~~. 
But I see you're saying that the example function invocation is missing it. 

> why expose delivery amount/units but not sent amount/units?

Interesting ? Are you proposing exposing both ? 

@sentientwaffle 
Btw, [CoilPreview](https://chrome.google.com/webstore/detail/coilpreview/iehmfkldnblennopinmmagfidpflefkp) has a lot of this implemented.

sublimator commented:
@sentientwaffle 
Agreed

sublimator commented on static/specification.html:
```diff
@@ -144,595 +204,516 @@ <h2>
+          Although a single HTML document can have multiple [^link^] elements
+          with the "monetization" link type, the user agent might restrict the
+          number of concurrent [[STREAM]] connections it can simultaneously
+          support. In such a case, an "error" event will be fired at the
```
Will the "error" event be of type "ErrorEvent" ?

sublimator replied:
Nevermind, I guess it's not really that suitable

sublimator replied:
Though window.onerror emits an ErrorEvent it seems

marcoscaceres replied:
ErrorEvent seems appropriate... let me check...

sublimator replied:
The nice thing about ErrorEvent is it supports communicating something more than just "error" via the "error: Error" member. But I wonder if it would be that appropriate when it comes to native implementations? It seems to be for "script" errors

sublimator replied:
> The ErrorEvent interface represents events providing information related to errors in scripts ** or in files. **
Or in (cpp) files :) ??

marcoscaceres replied:
Actually, no... it seems just regular errors are thrown, as ErrorEvent are for script errors.

sublimator replied:
Pity :/

marcoscaceres commented on static/specification.html:
```diff
@@ -144,595 +204,516 @@ <h2>
-              </ul>
+        <li>Let |eventInitDict:MonetizationEventInit| be a
+        {{MonetizationEventInit}} dictionary, whose members are initialized to
+        match |packet|'s details.
```
Here, I think we need to "canonicalize" the assetCode to uppercase and recommend ISO currency codes... basically, we need to copy/paste:
https://www.w3.org/TR/payment-request/#dom-paymentcurrencyamount-currency

marcoscaceres commented:
> §8, Is assetScale intentionally a string (instead of a integer/number)? (If so, why?)

I think it might be because the `assetScale` could get really small/large (beyond what could be safely represented in JS without rounding errors). So to avoid loss of precision, it's a string. It would then be up to the developer to safely convert it to number if it's less than some length, or use some other strategy to handle it.

marcoscaceres commented:
@sentientwaffle: 

> §1.1: The example shows onmonetization="sayThanks()" and function sayThanks(event) { — where does event come from?

Good point. Fixing. 

> Also, why expose delivery amount/units but not sent amount/units?

I think that would expose information about the sender... e.g., sending in AU dollars. That would violate user privacy. 

> §8.1, typo: The amount received as by an ILP packet.. Also, "delivered" might be more appropriate than "received", since this is mostly written from the perspective of the web browser, not the monetization receiver.

Fixed.

> §8.2, nit: replace "describing" with "identifying" or "indicating".

Fixed. 

> §9, typo: there's an extra double-quote after monetization.

Fixed.

> It would be good to either include or link to some more comprehensive examples, e.g. bypassing a paywall.

Added a better example of using events and verifying a receipt.

marcoscaceres commented:
@sublimator, @AlexLakatos, could you please review: https://github.com/WICG/webmonetization/pull/193/commits/3f66ff181c18fe8bb022ded508310937bd17363d

sublimator commented on static/specification.html:
```diff
@@ -110,16 +110,88 @@ <h3>
+                receipt: event.receipt
+              });
+
+              const response = await fetch('https://verifier.example/verify/', {
```
@wilsonianb 
Generally we'll have the paymentPointer encoded in the url here right ?

sublimator replied:
Maybe the example should have something with `encodeURIComponent(event.paymentPointer)`

marcoscaceres replied:
right, you mean if we send it as URL parameter? If yes:

```JS
const url = new URL('https://verifier.example/verify/');
url.searchParams.set("verify", event.paymentPointer);
fetch(url, ...);
```

sublimator replied:
yeah, just some way of encoding the pp into the url
tbh, I'm not sure if it's strictly needed

Brandon ? 
@wilsonianb

sentientwaffle commented:
> I think it might be because the assetScale could get really small/large (beyond what could be safely represented in JS without rounding errors). So to avoid loss of precision, it's a string. It would then be up to the developer to safely convert it to number if it's less than some length, or use some other strategy to handle it.

When paying in nanodollars, the scale is `9`, not `1e9`. `Number.MAX_SAFE_INTEGER === 9007199254740991` -- I think we can reasonably assume that asset scales won't be nearly that high. The smallest possible asset scale is `0`, i.e. it cannot be subdivided. I don't think there are any precision issues from requiring `assetScale` to be a (positive) JS integer.

marcoscaceres commented:
@sentientwaffle, I set it to "unsigned long long" (but maybe overkill, as JS can't represent that anyway):

> The unsigned long long type is an unsigned integer type that has values in the range [0, 18446744073709551615].

Maybe I should just set it to "unsigned long"?:

> The unsigned long type is an unsigned integer type that has values in the range [0, 4294967295].

sublimator commented:
STREAM encodes it as a uint8
https://github.com/interledgerjs/ilp-protocol-stream/blob/master/src/packet.ts#L274
https://github.com/interledger-rs/interledger-rs/blob/master/crates/interledger-stream/src/packet.rs#L622-L633

sentientwaffle commented:
I agree that a `uint8` is plenty, and matches up with ILP-STREAM.

sentientwaffle commented:
Should the `amount` be  a `bigint` instead of a `string`?

sublimator commented:
Yeah, good question! bigint probably does make more sense in 2022

sabinebertram commented on static/specification.html:
```diff
@@ -144,595 +312,516 @@ <h2>
-        Unlike [[[payment-Request]]] and the [[[Payment-Handler]]], Web
+        Unlike [[[Payment-Request]]] and the [[[Payment-Handler]]], Web
         Monetization is intended for continuous payments rather than discrete
         payments. Additionally, it is designed to <em>not</em> require any user
```
If tipping is WM, then WM is not just for continuous payments and also requires active user interaction. If tipping is not WM, we could still have WM via Open Payments, which would be recurring payments without user interaction.

marcoscaceres replied:
> If tipping is WM, then WM is not just for continuous payments and also requires active user interaction.

Right, the point we've been trying to make is that discreet and continuous payments are indistinguishable. They are just "payment" events that just happen to be of different monetary values (e.g, imagine a stream like this: .36c, .36c, .36c, $5, .36c, .36c, .36c,  $20... .36c,... and so on )

sabinebertram reviewed (commented):
I have just added some first thoughts. I'll write out the web monetization use case with Open Payments v2 now and then may add some more comments.

marcoscaceres commented on static/specification.html:
```diff
@@ -58,17 +58,17 @@
-        receive micro-payments from users, facilitated by the browser and a
-        user's Web Monetization provider.
+        Web Monetization allows websites to automatically receive payments from
+        users, facilitated by the user agent and a user's preferred
```
Generally speaking, the implementation being a "extension" is an implementation detail. For example, various parts of Firefox are implemented as "system extensions", but they are indistinguishable from any other part of the browser.

sabinebertram replied:
I see. Thanks for clarifying!

marcoscaceres commented on static/specification.html:
```diff
@@ -93,34 +93,202 @@
-          provider uses the [[[SPSP]]] (<abbr title=
-          "Simple Payment Setup Protocol">SPSP</abbr>) on top of the
+          The party making payments on behalf of the user. A monetization
+          provider leverages the [[[SPSP]]] (<abbr title=
```
No reason :) Should I use https://docs.openpayments.dev/ as the specification reference? 

Question: Given that SPSP and Open Payments are both Interledger specs, is open Payments a new(er) version of SPSP? Would a user agent need to support both?

sabinebertram replied:
> Should I use https://docs.openpayments.dev/ as the specification reference?

I think that is best. I think we'll update that first as soon as we are happy with v2.

> is open Payments a new(er) version of SPSP?

No. With the direction we are going with OPv2, it's rather another application layer protocol. They will probably both co-exist. As of now, Rafiki will implement both.

> Would a user agent need to support both?

For now, yes.

marcoscaceres replied:
Ok, we added a concept of a "payment connection" which abstracts away STREAM and OP.

marcoscaceres commented on static/specification.html:
```diff
@@ -144,595 +312,516 @@ <h2>
     </section>
```
Right, we probably want to expand this out to say that the amounts can vary. Tips would still be part of a continuous stream, even if the stream only contains a one off payment event. 

So, how about:

> Users retains control of if, when, and how payments are made to the site. For example, micropayments of a predetermined monetary value could be "streamed" to the site over time. Alternatively, the user could make a discreet "one-off" payment of any arbitrary monetary amount, even while micropayments are simultaneously being streamed.

sabinebertram replied:
I like the suggestion. The one thing I would change is 

> Alternatively, the user could make **~~a~~ one or many** discreet "one-off" **payment(s)** of any arbitrary monetary amount, even while micropayments are simultaneously being streamed.

This includes tipping as well as recurring discreet payments via OP.

marcoscaceres commented on static/specification.html:
```diff
@@ -144,595 +312,516 @@ <h2>
+          <li>
+            <p>
+              If the user agent has exhausted the number of allowed [[STREAM]]
+              connections, set |success| to false.
```
Ok, I wonder if we should just allow for a more generalized fail here... "has exhausted resources" or something.

marcoscaceres commented:
Ok, all outstanding comments have been dealt with.

sabinebertram commented on static/specification.html:
```diff
@@ -93,34 +106,208 @@
-          [[[Interledger]]] (<abbr title=
-          "Interledger Protocol">ILP</abbr>)/[[[STREAM]]] to provide a
-          high-level way to send money and data.
+          The party making payments on behalf of the user. A monetization
+          provider leverages the [[[Interledger]]] suite of protocols and
+          technologies (e.g., [[Open-Payments]] or [[STREAM]]) to provide a
+          high-level way to pay a [=monetization receiver=].
```
Mentioning Open Payments and STREAM in the same breath may be confusing since OP is the application layer and STREAM is the transport layer. 

Maybe say
> A monetization provider leverages the Interledger Protocol suite of protocols and technologies (e.g., [Open-Payments] or [SPSP] , both based on [STREAM]) to provide a high-level way to pay a monetization receiver.

marcoscaceres replied:
Love it! Thank you.

sabinebertram commented on static/specification.html:
```diff
@@ -93,34 +106,208 @@
+          <dfn>Payment connection</dfn>
+        </dt>
+        <dd>
+          A network connection made over the [[[Interledger]]] network using,
+          for example, [[Open-Payments]] or [[STREAM]], for the purpose of
+          facilitating a payment.
```
same as above

However, I feel like it should not just be 
> A network connection made over the Interledger Protocol network using, for example, [Open-Payments] or [SPSP], both based on [STREAM], for the purpose of facilitating a payment.

I'm hesitating because 
* when using SPSP, a STREAM connection is actually opened by the agent
* when using OP, the agent will create an outgoing payment at a wallet, which then will open the STREAM connection

Maybe "payment connection" is not the right term. I'll think about that a bit more. 

Maybe @adrianhopebailie has an idea how to rephrase that to make it clearer.

AlexLakatos replied:
The point I wanted to make here was, from the point of view of a browser: "Web Monetization over HTTP (OP)" vs "Web Monetization over STREAM". The Web Monetization Agent in the browser decides which transport layer to use to make that connection in order to start payments. "Payment connection" was the best description we could come up with, not the ideal term. Looking forward to a better term.

marcoscaceres replied:
Ok, happy to change this... everyone, give this a think and let's see if we can come up with something better for when we meet next week! ⚗️

marcoscaceres replied:
Oh! I REALLY like the precision, @sabinebertram! Just to unpack it, I split the conditionals into an unordered list. 

What do you think?

sabinebertram replied:
Perfect.

sabinebertram commented on static/specification.html:
```diff
@@ -138,605 +325,535 @@ <h2>
+        <li>Users retains control of if, when, and how payments are made to the
+        site. For example, micropayments of a predetermined monetary value
+        could be "streamed" to the site over time. Alternatively, the user
+        could make a one or many discreet "one-off" payment(s) of any arbitrary
```
```suggestion
        could make one or many discreet "one-off" payment(s) of any arbitrary
```

sabinebertram reviewed (changes requested):
This is shaping up nicely!

Most of these are just typos 
(or what I believe to be typos. Don't take my word for granted, I'm not a native speaker)

sublimator commented on static/specification.html:
```diff
@@ -93,34 +109,209 @@
+            &lt;script&gt;
+              const link = document.querySelector('link[rel="monetization"]');
+              link.addEventListener("monetization", event =&gt; {
+                // See how much your received and in what currency
```
your -> you

sublimator reviewed (changes requested):
Can never remember how to "suggest changes"

marcoscaceres commented:
@sublimator, 
> Can never remember how to "suggest changes"

Yeah, it's confusing. You need to click on "files changed" tab a the top, and then select the lines you want to change on the right:
 
<img width="135" alt="Screen Shot 2022-03-04 at 12 36 43 am" src="https://user-images.githubusercontent.com/870154/156575464-a9ea7ffe-4a48-4605-88e1-2717fb1929ca.png">

then press this little "+-" button when the box appears:

<img width="391" alt="Screen Shot 2022-03-04 at 12 34 27 am" src="https://user-images.githubusercontent.com/870154/156575150-9d8f24dd-a548-4b8b-b00c-f2069ab75eeb.png">

sublimator commented on static/specification.html:
```diff
@@ -819,9 +825,10 @@ <h2>
-        RECOMMENDED that developers:
+        Uphold), a <abbr title="Cross Site Scripting">XSS</abbr> attack could
+        inject a malicious [=payment end-points=] into a page that uses the
+        same service. To mitigate such an attack, it RECOMMENDED that
```
it is

sublimator commented on static/specification.html:
```diff
@@ -819,9 +825,10 @@ <h2>
-        into a page that uses the same service. To mitigate such an attack, it
-        RECOMMENDED that developers:
+        Uphold), a <abbr title="Cross Site Scripting">XSS</abbr> attack could
+        inject a malicious [=payment end-points=] into a page that uses the
```
singular/plural tension

marcoscaceres commented on static/specification.html:
```diff
@@ -138,605 +335,536 @@ <h2>
+        As [=payment end-points=] are generally provided as a service (e.g.,
+        Uphold), a <abbr title="Cross Site Scripting">XSS</abbr> attack could
+        inject a malicious [=payment end-points=] into a page that uses the
+        same service. To mitigate such an attack, it RECOMMENDED that
```
```suggestion
        same service. To mitigate such an attack, it is RECOMMENDED that
```

marcoscaceres commented on static/specification.html:
```diff
@@ -138,605 +335,536 @@ <h2>
+      <p>
+        As [=payment end-points=] are generally provided as a service (e.g.,
+        Uphold), a <abbr title="Cross Site Scripting">XSS</abbr> attack could
+        inject a malicious [=payment end-points=] into a page that uses the
```
```suggestion
        inject malicious [=payment end-points=] into a page that uses the
```

marcoscaceres commented:
Thanks again @sublimator! fixed the typos. If you are happy with the current state, can you give it a ✅?

sublimator commented:
> If you are happy with the current state, can you give it a ✅?
Overall I am happy with this, but would like to kick the tires more on this before officially giving a ✅
At the moment I'm sidetracked with trying to get puppeteer working with Firefox so I can craft more tests. 
After that I would like to build some examples myself to get a a better feel for the API.

marcoscaceres commented:
> At the moment I'm sidetracked with trying to get puppeteer working with Firefox so I can craft more tests.

If you do come up wit tests, we should try to put things up at https://github.com/web-platform-tests/wpt

sublimator commented on static/specification.html:
```diff
@@ -138,605 +335,536 @@ <h2>
+          with the "monetization" link type, the user agent might restrict the
+          number of concurrent [=payment connections=] it can simultaneously
+          support. In such a case, an "error" event will be fired at the
+          [^link^] element where a connection can't be established.
```
Without any real detail on the error event how would a web application frontend distinguish this (connection limit exceeded) from an SPSP error?

marcoscaceres replied:
I guess the question (for those more knowledgeable in this space) is if developers can do anything meaningful with the distinction? We couldn’t originally come up with anything so I’m interested in hearing use cases.

sublimator replied:
Well let's start by listing the current errors:
1. SPSP request failed
2. SPSP JSON malformed
3. Connection limit exceeded 
4. ?  What else am I missing

I guess I'm wondering about user feedback for 3. What would it look like to them? 
It's just "not working"

sublimator replied:
Continuing on with this. There was a time in the evolution of the spec where the "error" event bubbled up the tree, but then it occurred to author that that would be problematic as there would be an unintended consequence of it triggering peoples handlers for window etc. 

There was also some discussion as to whether ErrorEvent might be more suitable as it had an `error` field where more data would be attached. It was found not to be suitable. 

We then went back to simply Event('error'). 

ppk suggested we at the least log error information to the console when localStorage.WM_DEBUG was set 

However, it's possible we could use a MonetizationError Event subclass with type 'monetizationerror' which we could actually endow with some useful information.  It could also bubble up the tree like the monetization events.

sublimator replied:
@AlexLakatos 

Can you speak to this, given you've effectively inherited the PR ?

The overview comment includes:
- Developer-recoverable errors are handled via the <link onerror> event handler.

The spec itself talks nothing of what errors are classed as developer recoverable, and which are non-recoverable. 
My recollection is quite fuzzy, but I gathered from somewhere it means that the payment pointer can possibly be changed to remedy the 'error' state ? 

> Without any real detail on the error event how would a web application frontend distinguish this (connection limit exceeded) from an SPSP error?

^

AlexLakatos replied:
The only action for every error that's "recoverable" from a developer's point of view is to change the payment pointer on the page. 

We thought adding extra information to the error won't help in this case, and a custom event, or even ErrorEvent wouldn't help here. A simple Error event would be better here.

sublimator replied:
https://github.com/WICG/webmonetization/pull/193#pullrequestreview-1215158725

adrianhopebailie commented:
@marcoscaceres , you said above, with regard to exposing the sent amount

> I think that would expose information about the sender... e.g., sending in AU dollars. That would violate user privacy.

Is your contention that exposing the currency used is a privacy violation? If so I want to push back on that a little for both technical and UX reasons.

On the technical side, we'd love to deprecate "receipts" as they're very difficult to implement and arguably are not that useful (per packet). If the model of integration we are moving toward is adopted (ILP-capable wallets expose standard APIs - called the [Open Payments APIs ](https://docs.openpayments.guide) and WM Providers use those to send payments on behalf of users) then surfacing receipts is very challenging.

Assuming a setup like this:
User with WM Provider (Coil) setup in their browser. WM Provider has access tokens to create outgoing payments from the users wallet (Uphold) via the Open Payments APIs.

1. User visits a website and the WM Provider is passed the Payment Pointer that was extracted from a link tag on the page.
2. WM Provider decides it should send 50c to the website at that Payment Pointer.
3. WM Provider calls out to the wallet to [create an outgoing payment](https://docs.openpayments.guide/reference/create-outgoing-payment) where the `receivingAccount` is the Payment Pointer and the `sendAmount` is 50c.
4. The wallet calls out to the receiving wallet (where the website has its account) and [creates an incoming payment](https://docs.openpayments.guide/reference/create-incoming-payment) with no fixed amount but it get's back a new unique URL for that resource.
5. The wallet updates the outgoing payment resource's `receivingPayment` property to be equal to the URL of the new payment it has just created. _(There is an exception condition if the receiver is a legacy wallet and only supports SPSP, I'll describe that later)_
6. The WM Provider emits the monetization event with the amount of 50c and the URL of the incoming payment at the receiving wallet as a new property. There is no receipt property.
7. The website can choose to:
  7.1 Trust the WM Provider has sent the payment and act accordingly based on an amount sent of 50c
  7.2 Poll the incoming payment to determine the amount received

This seems the best balance for UX and privacy too. 

- The website will get the monetisation event sooner (as soon as the payment is authorised but not waiting for it to be sent and confirmed).
- The website can be optimistic and assume it has been paid but also asynchronously verify this and how much it actually received.
- The WM Provider can emit the amount in whatever currency it wishes based on user preference (e.g. if it sends AUD it could apply a conversion and emit an event in USD) as this is just an indicative value for the website.
- The website can provide UX that makes sense to the user irrespective of the fees and exchange rate that might be applied (i.e. The user sent USD 0.50 so a thank you message of "thanks for the 872365 XRP tip" would be kind of weird).

adrianhopebailie commented:
Also note, part of the rationale for this flow is that we expect there to be a quite high trust between wallets so it should be straight-forward for a wallet to create an incoming payment at another wallet (step 4 above).

If the URL of that incoming payment is given to the account holder then they will also have permission to query this URL (step 7.2 above).

adrianhopebailie commented:
> There is an exception condition if the receiver is a legacy wallet and only supports SPSP, I'll describe that... **below**

If the sending wallet attempts to create an incoming payment at the receiving wallet and this fails because the receiving wallet only supports SPSP and not Open Payments then it sets the value of the `receivingPayment` to the same value as `receivingAccount`.

In this case the WM Provider knows that there is no incoming payment URL it can pass to the website and the webmonetization event has neither a receipt or an incoming payment URL.

The incentive is with websites to use wallets that support Open Payments if they need to verify their incoming payments.

adrianhopebailie commented:
I have tweaked this flow in my PR at #269. (Specifically noting it is the UA that would create the outgoing payments, not the WM Provider).

> Now, for most of the world, your wallet is not really a choice. Websites that want to get monetized in most countries dont't have alternatives or a choice of wallets. We should design for that.

I think it would be good to consult the existing wallets on what it would take for them to support, at a minimum **incoming payments**. My guess is it would be easier than receipts and we can try to align timelines for when wallets support this and when the new spec is published?

We COULD try to find a way to bubble up receipts to the UA but I am conscious of note corrupting the Open Payments API design to accommodate a feature that we know we want to deprecate.

marcoscaceres commented:
Thanks for the clarification @adrianhopebailie. That makes sense... and yeah, if receipts can go away, all the better.

sublimator commented on static/specification.html:
```diff
@@ -93,34 +105,220 @@
+              link.addEventListener("monetization", event =&gt; {
+                // See how much was sent and in what currency
+                const { amount, assetCode, assetScale } = event;
+                console.log(`Browser sent ${assetCode}${amount / (10 * assetScale)}.`);
```
This will throw `Uncaught TypeError: Cannot mix BigInt and other types, use explicit conversions` because we changed amount to a BigInt 

cc @Danisco212

sublimator replied:
Also, `amount / Math.pow(10, assetScale)`

sublimator commented on static/specification.html:
```diff
@@ -93,34 +105,220 @@
+
+              if (response.ok) {
+                // The incoming payment was fetched successfully
+                const { receivedAmount } = JSON.parse(response.json())
```
TODO: response.json() returns a promise

sublimator commented on static/specification.html:
```diff
@@ -93,34 +105,220 @@
+                // The incoming payment was fetched successfully
+                const { receivedAmount } = JSON.parse(response.json())
+                const { amount, assetCode, assetScale } = receivedAmount;
+                console.log(`Received ${assetCode}${amount / (10 * assetScale)}.`);
```
TODO: will fail due to not all operands being BigInt

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+        monetization event
+      </h2>
+      <p>
+        When the [=payment session=] has sent an |outgoing payment| with a
```
This means only one event will be fired, not upon each packet ? 
Or one tag configuration could fire multiple outgoing payments? 
Switch tabs and it pauses. 
Return to the tab 30 minutes later and it perhaps uses a new outgoing payment? 
It's a bit unclear to me where the payment is happening, from an OP server, or ILP running directly in the browser. 

Links to relevant OP specs for referenced concepts might be good (is there one in this doc already) ?

sublimator replied:
@adrianhopebailie
A monetization event per outgoing_payment rather than per packet ? 

Probably a question for you

adrianhopebailie replied:
This should be per **payment**, whatever that means. I believe that it should depend on the WM Provider.

i.e. If the provider sends 5 small packets and then emits 1 event to the browser then the website only gets one event. If the provider emits 5 events then the browser should emit 5 events IMHO.

sublimator replied:
> It's a bit unclear to me where the payment is happening, from an OP server, or ILP running directly in the browser.

In @AlexLakatos  model, as far as I can tell, the browser creates the STREAM itself, having been given a connection url/rate/token from the Provider component. (It's unclear to me how it can usefully give a rate based on just connection details to keep with the privacy model).

@adrianhopebailie 
It seems to me, in your model, the WM Provider could create a STREAM itself, in which case, it would likely emit an event for each packet, but it could also be sending payments via http requests, in which case there would just be one packet unless it polls, in order to support progress events. 

A website showing a counter (debatable how much value that really has, but let's just say a website frontend wants the events for some reason) would have to poll the incomingPayment url for updates, if it can't somewhat reliably assume that it will get progress events. 

Could OP support subscriptions (server push, websockets etc) at an incoming payment ?

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+        [SecureContext, Exposed=Window]
+        interface MonetizationEvent : Event {
+          constructor(DOMString type, MonetizationEventInit eventInitDict);
+          readonly attribute DOMString amount;
```
The amount of confusion/bugs we've had over the years forgetting whether the amount was a source or received amount. If we are to have an amount (not sure how it relates to an outgoing payment ), perhaps we should explicitly name it sourceAmount/destinationAmount etc. 

Why not both? 

I believe @adrianhopebailie  was arguing for sourceAmount only, so that the monetization event could be reported ASAP. If you reported both, you’d have to wait for the packet to be sent to get the (received)Amount.

Why not sourceAmount only?

As a website, you own the payment pointer, you know the destination currency, and you want to code around that.

adrianhopebailie replied:
I argue that the purpose of the event is to **notify the website that a payment has been sent**, not that a payment has been received.

Assuming that we are able to notify the website that a payment has been received seems leaky and error prone (as we've discovered). It makes it very difficult to be flexible in the implementation of the payments behind this API.

Ultimately the website is going to do 1 of 2 things with this information:

1. Trust that the amount was sent since its so small and the good or service provided in return so low in value that it doesn't bother verifying the payment was received. In this case it either does a rudimentary conversion to an asset it knows, using some public exchange rate, or simply doesn't bother because it's not checking it received the funds anyway so what value does it get from converting the send amount to the receive currency.

2. It verifies the payment at the receive account and in so doing gets the receive amount in the currency of the receive account.

I suggest we stick with sendAmount in the event and leave it to the website to calculate or fetch the receive amount

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
-            <code>meta.content = meta.content</code>) new <code>&lt;meta
-            name="monetization"&gt;</code> tag seen for the life of the page:
+        <li>Let |eventInitDict:MonetizationEventInit| be a
+        {{MonetizationEventInit}} dictionary, whose members are initialized to
```
Elsewhere it says:
```
 The {{MonetizationEvent}} DOM events provide information such as the
          amount sent, the currency of the payment (what we call the "asset
          code"), and a link to the incoming payment at the [=monetization
          receiver=]
```          

The outgoing_payment references an incoming_payment to be attached to the event ?

sublimator commented on static/specification.html:
```diff
@@ -93,34 +105,220 @@
+              if (response.ok) {
+                // The incoming payment was fetched successfully
+                const { receivedAmount } = JSON.parse(response.json())
+                const { amount, assetCode, assetScale } = receivedAmount;
```
This could effect #240

wilsonianb replied:
> Does Open Payments express amount values as amount/assetScale ?

```suggestion
                const { value, assetCode, assetScale } = receivedAmount;
```
https://docs.openpayments.guide/reference/get-incoming-payment
https://github.com/interledger/open-payments/blob/4c873dba89164decffbe84905d12f1d4ec045389/open-api-spec.yaml#L1159-L1195

sublimator replied:
@wilsonianb 
Thanks

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
-            (i.e. not inside an iframe)
-            </li>
-          </ul>
+        <li>If |target| is `null`, then return.
```
From https://github.com/WICG/webmonetization/issues/19#issuecomment-705407129
<img width="871" alt="image" src="https://user-images.githubusercontent.com/525211/194221337-9e709c66-7a7c-495d-8da5-b7bb3560a838.png">

adrianhopebailie replied:
Good question. I think it makes sense for browsers to support link headers but we'd need to define how the events are handled in the DOM.

sublimator replied:
Current examples also include checks for monetization tags (for whatever purpose)

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+        Further, the user agent gets the payment information from the [=payment
+        end-point=] to establish the [=payment session=]. This also ensures the
+        [=monetization provider=] doesn't have access to a user's browsing
+        history or to the [=payment end-point=].
```
It seems both outgoing and incoming payments are POST created resources 
https://docs.openpayments.guide/reference/create-outgoing-payment

Can a link's GET request suffice to get Open Payment connection details? 
Or some other step that requires the payment pointer (and perhaps some other credentials) is necessary ?

adrianhopebailie replied:
The privacy here is achieved by separating the processing of the payment pointer and the sending of the payment.

The user agent is the only thing that should process the payment pointer.

The WM Provider needs to send the payment and provide data back to the browser about when it is sending a payment, how much it sent and also how the website can verify the payment.

The interface between the browser and the WM Provider is not defined in this spec. How a browser chooses to allow WM Providers to integrate is up to them. Ideally it would be through the extensions system but it could be something else like the Payment Handler API.

OP was designed the way it is (create an incoming payment first) specifically to support this privacy-enhancing separation. i.e. The browser creates the incoming payment and only provides the connection URL to the WM Provider.

A browser does need to authenticate to create an incoming payment but it can dynamically register (i.e. no user interaction required) as a client with the wallet that issued the payment pointer which is a once off to prevent DDoS attacks.

sublimator commented on static/specification.html:
```diff
@@ -93,34 +105,220 @@
+              link.addEventListener("monetization", event =&gt; {
+                // See how much was sent and in what currency
+                const { amount, assetCode, assetScale } = event;
+                console.log(`Browser sent ${assetCode}${amount / (10 * assetScale)}.`);
```
> We are also training people with these examples to use the "string" typed amount as a number,

From https://webmonetization.org/docs/counter:
> Even though ev.detail.amount is a string that can represent a number up to 64 bits long in theory (too big for a JavaScript number), it's OK for us to convert it to a JavaScript number here. The micropayment amounts in Web Monetization are far from pushing the limits. With USD and a scale of 9, for instance, it would take around 500,000 years to overflow this number.

I believe STREAM [encodes](https://github.com/interledger/rfcs/blob/master/asn1/Stream.asn) amounts as varInt. 
TigerBeetle encodes transaction amounts as Uint64. 

If the packets are in practice small enough to be "OK", perhaps that should be enforced somewhere?

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+              sessions=], set |success| to false.
+            </p>
+          </li>
+          <li>Otherwise, establish a new [=payment session=].
```
Some breakdown here:
https://webmonetization-preview.netlify.app/docs/web-monetization-api#web-monetization-concepts-and-usage

sublimator replied:
> The Web Monetization Agent is designed to be implemented in browsers and expose the Web Monetization API. It's responsible to:
get the rate, token, and payment connection address from the Web Monetization Provider

How will tips be supported in this model? 
How is the rate calculated ? It's a flat rate? If so, the backend could set this. 
Would the provider need access to the active tab' url to create a dynamic rate?

adrianhopebailie replied:
Only the browser implements this spec. Its role is to provide an interface for websites to code against.

The browser is responsible for fetching the payment pointer URL and doing any processing necessary to pass payment details to the WM provider without leaking any details of the website that has been visited or the payment pointer. 

Following that, the website emits a load event so the website knows that WM is enabled and the payment pointer was successfully processed. What the website does with that info is up to it. There is no guarantee that any payments will follow or if they do, how often, or how large they will be.

The last role of the browser is to emit an event each time a payment is sent. The event should provide enough data for the website to verify the payment independently since the event is simply a notification that there was an attempt to pay. Again, depending on the context, the website may do different things with this information. For micropayments in exchange for streaming content the website may not bother verifying all payments for larger payments in exchange for some good or service the website might verify the payment before delivering the good or service.

The WM Provider is responsible for executing payments using the details provided by the browser. The WM Provider defines how much to send and when either through some heuristics or user input (e.g. tipping).

The interface between the browser and WM Provider is NOT defined here. This is intentional so that we leave flexibility in how this might evolve in future. If we hadn't leaked the receipt logic out through this API we might have been able to easily switch between SPSP and OP without any changes :(

sublimator commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+              <code>link</code> element that is already <span>browsing-context
+              connected</span>.
+            </p>
+          </li><!-- e.g. rel="" changed, href="" set... -->
```
### States

One of the WM2 drafts had states as so (definitions my own)

#### * idle
The link has not been processed

Payments can not come

#### * interactive
Web Monetization is enabled, provider configured, and the link has been 
processed.

Payments /may/ come

(This has been referred to as the payment pointer being validated.
It's unclear if that meant making a request, or simply checking if
it was well formed, though I believe it was the latter)

From [this comment](https://github.com/WICG/webmonetization/issues/45#issuecomment-983013053)

    We've changed the flow for the state of "monetized" in the current
    iteration of the spec. The "interactive" state triggers as soon as
    we validate the payment pointer, hence providing a middle solution
    here. 

    We're showing content as soon as the first payment can be made
    (which is similar to "you have a web monetization provider"),
    instead of showing it only after the first payment has been made.

#### * error

A non recoverable error has occurred which can only be remedied by
reconfiguring the link tag (e.g. new href)

Payments can not come

#### Valid transitions:
* idle -> error
* idle -> interactive
* interactive -> error

#### Link events with implied state

Upon moving to multiple link tags, the state is implied by events:
Link state starts in idle

#### * 'load' event
transitions state to interactive

#### * 'error' event
transitions state to error 

Per tag configuration there is only ever one load and/or error event, so it
fits with the previous model's valid transitions.

In the move from explicit state to implied, it seems interactive has changed
definition from simply validating the payment pointer (check it's well
formed) to actually requesting it.

Given 'load' does not actually map to how the "loading" of a payment pointer is
actually done ([1]) then why not emit 'load' as soon as the payment pointer has
been deemed well formed. If the SPSP (or OP equivalent) fails, then transition
to the error state by emitting an error. 

Most of the time a configured payment pointer will be valid, and most of the
time the payment pointer will be accessible via network. Why delay the
interactive state for edge cases?

[1] Where it may be loaded more than once, or actually not at all via the
link network machinery, but rather by some other component, or at a WM provider
backend. 

### Where this model may break down

Automatic payment is rejected due to disabling

####  You don't emit the load event

State is not 'interactive'

'monetization' event may still come from user initiated tip, even
 though this breaks the state model, where it says no payments will
 come when the link is in the idle state

[Example](https://webmonetization-preview.netlify.app/docs/exclusive-content#code) using only load event to unlock exclusive content

#### You do emit the load event, to keep state coherent

In this case automatic payments are disabled

This means that the load event is not as useful for guessing whether
payments will come, and "optimistically" show content.

adrianhopebailie replied:
> How does disabled monetization fit into this ? Just load anyway, and emit 'load' event?

No. If monetization is disabled, don't emit an event

> What about requiring connection details more than once per tag configuration ?

Getting the connection details a second time shouldn't emit a new load event. The load event is related to processing the payment pointer.

I realise this is problematic when trying to do client-side accounting using receipts but this is yet another reason why we want to deprecate receipts.

sublimator replied:
> The load event is related to processing the payment pointer.

In the spirit of optimism, can we emit it earlier than having actually made a network request ?

> Given 'load' does not actually map to how the "loading" of a payment pointer is
actually done ([1]) then why not emit 'load' as soon as the payment pointer has
been deemed well formed. If the SPSP (or OP equivalent) fails, then transition
to the error state by emitting an error.

^

adrianhopebailie replied:
I think the outstanding question here is what to do in the case of automatic streaming being disabled by the user or provider but then the user sends a payment explicitly.

Does the browser emit a load event and then a monetization event or just a monetization event?
I suggest we get some implementation experience with tips before we try to make a final decision.

AlexLakatos commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+        Unlike [[[Payment-Request]]] and the [[[Payment-Handler]]], which only
+        supports "one-off" payments, Web Monetization provides a [=payment
+        session=] that supports both continuous payments and "one-off"
+        payments.
```
I partially agree, it's not about user interaction, but about user initiation. Users are not expected to initiate the payment in WM, as opposed to Payment Request. I'll update the spec to reflect that, nice catch!

sublimator replied:
Ack

AlexLakatos commented on static/specification.html:
```diff
@@ -138,605 +336,555 @@ <h2>
+          </li>
+        </ul>
+        <p>
+          A user agent MUST NOT <span>delay the load event</span> for this link
```
It shouldn't be easy for a website to figure out a user is able to do web monetization, but has chosen not to do so. We think the event shouldn't fire when monetization is disabled on a page. And that's because the user agent shouldn't process the payment pointer in the first place when monetization is disabled on a page.

sublimator replied:
> It shouldn't be easy for a website to figure out a user is able to do web monetization, but has chosen not to do so.

User comes across site with automatic "zero friction" WM enabled 
User Agent sends packets of money
User decides they don't want to support the site and disables automatic WM for the site
Site sees packets on initial page loads session (cookies/whatever ) 
Site doesn't see packets on subsequent page loads  

It's not going to be that hard 

Given the tools (we've never offered them previously) there could be a subset of WM users who disable automatic WM globally and only tip. 

Are we saying when automatic payment is disabled, the load event doesn't fire?

adrianhopebailie replied:
I suggest we leave this to UA vendors to figure out based on their own privacy policies and implementation experience. They may provide us with extra detail we need to add to the spec.

sublimator commented on static/specification.html:
```diff
@@ -93,34 +105,220 @@
+          <dfn>Payment session</dfn>
+        </dt>
+        <dd>
+          An session between a [=monetization provider=] and [=monetization
```
What are your thoughts on sessions lasting beyond a tag configuration (page lifetime at most) ? 
Consider a user that has automatic payments turned off, and only likes to tip. 
He tips $5, but then on reloads he's presented with ads. 
Or on a mobile user with an agent that likes to pay big amounts periodically. 
Leave the session to be created using cookies and stitching together a heap of incoming payments / outgoing payments ?

sabineschaller commented on static/specification.html:
```diff
@@ -93,34 +106,220 @@
+        </h3>
+        <p>
+          The {{MonetizationEvent}} DOM events provide information such as the
+          amount sent, the currency of the payment (what we call the "asset
+          code"), and a link to the incoming payment at the [=monetization
```
since we are deprecating `assetCode`, we should only mention `amountSent` here.

sabineschaller commented on static/specification.html:
```diff
@@ -93,34 +106,220 @@
+                incomingPayment: event.incomingPayment
+              });
+
+              const response = await fetch('/verify', {
```
I'm not sure if `/verify` is a good example since it's not part of the Open Payments spec.

adrianhopebailie replied:
I have changed this example to simply query the incoming payment.

sabineschaller commented on static/specification.html:
```diff
@@ -93,34 +106,220 @@
+          <dfn>Payment Pointer</dfn>:
+        </dt>
+        <dd>
+          A [=URL=] to an <a data-cite="Open Payments#specification">Open
+          Payments API entry-point</a> (i.e., a JSON resource containing
```
Is it really just an Open Payments Payment Pointer? Should we include a warning that SPSP is deprecated?

adrianhopebailie replied:
I'm not sure it's needed. I wouldn't put it here specifically if we do add it.

sabineschaller replied:
Maybe just "API entry-point", dropping the "Open Payments" part?

sabineschaller replied:
@adrianhopebailie 👆

adrianhopebailie replied:
Why omit Open Payments? 
If this is just an "API entry-point" then how is an implementation supposed to know what API is implemented?

sabineschaller replied:
I guess I'm thinking about backwards compatibility with SPSP. What if the account provider only supports SPSP? Then a GET to the payment pointer is redirected to `https://payment.pointer/alice/.well-known/pay`and it doesn't respond with the Open Payments json blob or the acts as entry point to a grant request.

sabineschaller replied:
Since there is no mentioning of SPSP in the spec anymore, I guess we can keep it as is.

sabineschaller commented on static/specification.html:
```diff
@@ -138,605 +337,612 @@ <h2>
+      <section>
         <h2>
-          Events
+          <dfn>receipt</dfn> attribute
```
Is missing (deprecated)

sabineschaller commented on static/specification.html:
```diff
@@ -138,605 +337,612 @@ <h2>
-          DEPRECATED: The `Web-Monetization-Id` header is deprecated in favor
-          of receipts.
+        <p>
+          A [=URL=] representing the [=payment pointer=] that has been
+          monetized. When getting, returns the value it was initialized with.
```
I don't get what this is saying

sublimator replied:
it's a getter, and returns the value passed in, in the constructor 
overly wordy imo

sublimator reviewed (dismissed):
tag it

sabineschaller commented:
@AlexLakatos can you redeploy to netlify for a final review?

AlexLakatos commented:
Here it is, now with the respec error fixed as well: https://webmonetization-preview.netlify.app/specification.html

sublimator commented:
>>> Should a load event be emitted before a monetization event if the browser is not streaming payments in the background but then the user sends a tip?

Or, perhaps more to the point: should be there a load event at all? and if not, perhaps changes circa incomingPayment id creation/aliasing to support instant monetization event emission? But then when you consider header links, where do you put the monetization event? window! Easy, right? But what about errors? Again, how do you determine what kind of error? Does window not already have error events? Yes, it does (so you might need a distinct error event)

Orienting things around a link api, similar to load/error in stylesheets, while on the surface, elegant, is not actually that flexible.
